# **ANOVA Test**
> ANOVA is a statistical test that stands for analysis of variance.

- ANOVA was developed by statistician and evolutionary biologist Ronald Fisher. The idea behind ANOVA is to compare different groups of samples to determine whether there is a significant difference between the groups.
- ANOVA is an extension of the t and the z test and was developed to compare more than two groups.
- The null hypothesis of ANOVA is that there is no difference between the groups. The alternative hypothesis is that there is a difference between the groups.
- ANOVA is an omnibus test, meaning it tests the data as a whole. In other words, it does not tell you which specific groups were significantly different from each other; it only tells you that at least two groups were different.

### **Types of ANOVA:**
There are three main types of ANOVA:
1. One-way ANOVA
2. Two-way ANOVA
3. N-way ANOVA

#### **1. One-way ANOVA:**
- One-way ANOVA is used to compare two or more groups of samples across one continuous independent variable.
- For example, you could use a one-way ANOVA to compare the height of people living in different cities.

#### **2. Two-way ANOVA:**
- Two-way ANOVA is used to compare two or more groups of samples across two independent variables.
- For example, you could use a two-way ANOVA to compare the height of people living in different cities and different countries.

#### **3. N-way ANOVA:**
- N-way ANOVA is used to compare two or more groups of samples across N independent variables.

### **Assumptions of ANOVA:**

- ANOVA has three main assumptions:
1. The samples are independent.
2. The samples are normally distributed.
3. The variance of each group is equal.

If these assumptions are not met, you may not be able to trust the results of your ANOVA.

## **1. ONE-WAY ANOVA**

### **1.1 One-Way ANOVA (Sample Data 1):**

In [1]:
import scipy.stats as stats

# Sample data

a = [2.3, 3.4, 4.5, 2.3, 3.4]
b = [11.2, 12.2, 13.2, 12.2, 12.3]
c = [22.3, 23.4, 24.5, 22.3, 23.4]

# perform one-way ANOVA
f_stat, p_val = stats.f_oneway(a, b, c)

# print the results base on if else conditions
if p_val > 0.05:
    print(f'p-value: {p_val}, Means are equal (fail to reject H0)')
else:
    print(f'p-value: {p_val}, Means are not equal (reject H0)')

p-value: 4.279979561060641e-13, Means are not equal (reject H0)


The above code is performing a one-way ANOVA (Analysis of Variance) on three groups of data to test if their means are equal. Here's a step-by-step explanation:

1. Import the `scipy.stats` module, which contains statistical functions.
2. Define three groups of data, `a`, `b`, and `c`.
3. Use the `f_oneway()` function from `scipy.stats` to perform a one-way ANOVA on the three groups. This function returns two values: the F statistic (`f_stat`) and the p-value (`p_val`).
4. Based on the p-value, make a decision about the null hypothesis. If the p-value is greater than 0.05, we fail to reject the null hypothesis and conclude that the means are equal. If the p-value is less than or equal to 0.05, we reject the null hypothesis and conclude that the means are not equal.

**Observations from the output:**

- The output shows that the `p-value` is approximately `4.28e-13`, which is significantly `less than 0.05`. 
- Therefore, we `reject the null hypothesis`. 
- This suggests that at least one group mean is statistically different from the others according to the one-way ANOVA.

### **1.2 One-Way ANOVA (Sample Data 2):**

In [5]:
import scipy.stats as stats

# Sample data: Growth of plants with three types of fertilizers
fertilizer1 = [20, 22, 19, 24, 25]
fertilizer2 = [28, 30, 27, 26, 29]
fertilizer3 = [18, 20, 22, 19, 24]

# Perform the one-way ANOVA
f_stat, p_val = stats.f_oneway(fertilizer1, fertilizer2, fertilizer3)

print("F-statistic:", f_stat)
print("p-value:", p_val)

# print the results based on if the p-value is less than 0.05

if p_val < 0.05:
    print("Reject null hypothesis: The means are not equal, as the p-value: {p_val} is less than 0.05")
else:
    print("Accept null hypothesis: The means are equal, as the p-value: {p_val} is greater than 0.05")

F-statistic: 15.662162162162158
p-value: 0.0004515404760997283
Reject null hypothesis: The means are not equal, as the p-value: {p_val} is less than 0.05


The above code is performing a one-way ANOVA (Analysis of Variance) test on the growth of plants with three different types of fertilizers. Here's a step-by-step explanation:

1. Import the necessary library: scipy.stats.
2. Define the sample data for the growth of plants with three types of fertilizers.
3. Perform the one-way ANOVA using `f_oneway()` function from scipy.stats. This function returns two values: the F-statistic and the p-value.
4. Print the F-statistic and the p-value.
5. Based on the p-value, a decision is made about the null hypothesis. If the p-value is less than 0.05, we reject the null hypothesis and conclude that at least one of the means of the plant growth for the three fertilizers is different. If the p-value is greater than or equal to 0.05, we fail to reject the null hypothesis and conclude that all the means are equal.

**Observations from the output:**

- The output shows that the `F-statistic` is approximately `15.66` and the `p-value` is approximately `0.00045`. 
- Therefore, we `reject the null hypothesis`. 
- This suggests that at least one of the means of the plant growth for the three fertilizers is `different` according to the one-way ANOVA test.

### **1.3 One-Way ANOVA using Statsmodels:**

In [2]:
# Install statsmodels
!pip install statsmodels --quiet

In [3]:
# Instaling libraries
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols

In [6]:
# Create a dataframe
df = pd.DataFrame({"fertilizer": ["fertilizer1"] * 5 + ["fertilizer2"] * 5 + ["fertilizer3"] * 5,
                   "growth": fertilizer1 + fertilizer2 + fertilizer3})
df.head()

fertilizer  growth
0  fertilizer1      20
1  fertilizer1      22
2  fertilizer1      19
3  fertilizer1      24
4  fertilizer1      25

In [7]:
# Checking for value counts of each fertilizer
df['fertilizer'].value_counts()

fertilizer
fertilizer1    5
fertilizer2    5
fertilizer3    5
Name: count, dtype: int64

In [8]:
# Fit the model using ols function
model = ols("growth ~ fertilizer", data=df).fit()

In [9]:
# Perform ANOVA and print the summary table
anova_table = sm.stats.anova_lm(model, typ=2)
print(anova_table)

                sum_sq    df          F    PR(>F)
fertilizer  154.533333   2.0  15.662162  0.000452
Residual     59.200000  12.0        NaN       NaN


In [10]:
# print the results based on if the p-value is less than 0.05
if anova_table["PR(>F)"][0] < 0.05:
    print("Reject null hypothesis: The means are not equal, as the p-value: {p_val} is less than 0.05")
else:
    print("Accept null hypothesis: The means are equal, as the p-value: {p_val} is greater than 0.05")


Reject null hypothesis: The means are not equal, as the p-value: {p_val} is less than 0.05


C:\Users\Hp\AppData\Local\Temp\ipykernel_252\2164794257.py:2: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  if anova_table["PR(>F)"][0] < 0.05:


In [11]:
# One-way ANOVA using statsmodels
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols

# Create a dataframe
df = pd.DataFrame({"fertilizer": ["fertilizer1"] * 5 + ["fertilizer2"] * 5 + ["fertilizer3"] * 5,
                   "growth": fertilizer1 + fertilizer2 + fertilizer3})

# Fit the model
model = ols("growth ~ fertilizer", data=df).fit()

# Perform ANOVA and print the summary table
anova_table = sm.stats.anova_lm(model, typ=2)
print(anova_table)

# print the results based on if the p-value is less than 0.05
if anova_table["PR(>F)"][0] < 0.05:
    print("Reject null hypothesis: The means are not equal, as the p-value: {p_val} is less than 0.05")
else:
    print("Accept null hypothesis: The means are equal, as the p-value: {p_val} is greater than 0.05")


                sum_sq    df          F    PR(>F)
fertilizer  154.533333   2.0  15.662162  0.000452
Residual     59.200000  12.0        NaN       NaN
Reject null hypothesis: The means are not equal, as the p-value: {p_val} is less than 0.05


C:\Users\Hp\AppData\Local\Temp\ipykernel_252\3735742722.py:18: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  if anova_table["PR(>F)"][0] < 0.05:


The above code is performing a one-way ANOVA (Analysis of Variance) test using the statsmodels library on the growth of plants with three different types of fertilizers. Here's a step-by-step explanation:

1. `Import the necessary libraries:` pandas, statsmodels.api, and statsmodels.formula.api.
2. `Define the sample data` for the growth of plants with three types of fertilizers and create a DataFrame.
3. `Fit the model` using the ols (Ordinary Least Squares) method from statsmodels.formula.api.
4. Perform the ANOVA using `anova_lm()` function from statsmodels.api and print the `summary table`.
5. Based on the p-value, a decision is made about the null hypothesis. If the p-value is less than 0.05, we reject the null hypothesis and conclude that at least one of the means of the plant growth for the three fertilizers is different. If the p-value is greater than or equal to 0.05, we fail to reject the null hypothesis and conclude that all the means are equal.

**Observations from the output:**

- The output shows that the `F-statistic` is approximately `15.66` and the `p-value` is approximately `0.00045`. 
- Therefore, we `reject the null hypothesis`. 
- This suggests that at least one of the means of the plant growth for the three fertilizers is `different` according to the one-way ANOVA test.

**Note:** The warning message indicates that using integer keys to access a value in a pandas Series is deprecated and will be removed in a future version. To access a value by position, you should use the `iloc` property. So, you should replace `anova_table["PR(>F)"][0]` with `anova_table["PR(>F)"].iloc[0]`.

## **2. TWO-WAY ANOVA**

In [12]:
# Sample data
data = pd.DataFrame({
    "Growth": [20, 22, 19, 24, 25, 28, 30, 27, 26, 29, 18, 20, 22, 19, 24,
               21, 23, 20, 25, 26, 29, 31, 28, 27, 30, 19, 21, 23, 20, 25],
    "Fertilizer": ["F1", "F1", "F1", "F1", "F1", "F2", "F2", "F2", "F2", "F2", 
                   "F3", "F3", "F3", "F3", "F3", "F1", "F1", "F1", "F1", "F1", 
                   "F2", "F2", "F2", "F2", "F2", "F3", "F3", "F3", "F3", "F3"],
    "Sunlight": ["High", "High", "High", "High", "High", "High", "High", "High", "High", "High", 
                 "High", "High", "High", "High", "High", "Low", "Low", "Low", "Low", "Low", 
                 "Low", "Low", "Low", "Low", "Low", "Low", "Low", "Low", "Low", "Low"]
})
data.head()

Growth Fertilizer Sunlight
0      20         F1     High
1      22         F1     High
2      19         F1     High
3      24         F1     High
4      25         F1     High

In [13]:
# Import Libraries
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols

# Sample data
data = pd.DataFrame({
    "Growth": [20, 22, 19, 24, 25, 28, 30, 27, 26, 29, 18, 20, 22, 19, 24,
               21, 23, 20, 25, 26, 29, 31, 28, 27, 30, 19, 21, 23, 20, 25],
    "Fertilizer": ["F1", "F1", "F1", "F1", "F1", "F2", "F2", "F2", "F2", "F2", 
                   "F3", "F3", "F3", "F3", "F3", "F1", "F1", "F1", "F1", "F1", 
                   "F2", "F2", "F2", "F2", "F2", "F3", "F3", "F3", "F3", "F3"],
    "Sunlight": ["High", "High", "High", "High", "High", "High", "High", "High", "High", "High", 
                 "High", "High", "High", "High", "High", "Low", "Low", "Low", "Low", "Low", 
                 "Low", "Low", "Low", "Low", "Low", "Low", "Low", "Low", "Low", "Low"]
})

# Perform two-way ANOVA
model = ols('Growth ~ C(Fertilizer) + C(Sunlight) + C(Fertilizer):C(Sunlight)', data=data).fit()
anova_table = sm.stats.anova_lm(model, typ=2)

print(anova_table)

# print the results based on if the p-value is less than 0.05

if anova_table["PR(>F)"][0] < 0.05:
    print("Reject null hypothesis: The means are not equal, as the p-value: {p_val} is less than 0.05")
else:
    print("Accept null hypothesis: The means are equal, as the p-value: {p_val} is greater than 0.05")

                                 sum_sq    df             F        PR(>F)
C(Fertilizer)              3.090667e+02   2.0  3.132432e+01  2.038888e-07
C(Sunlight)                7.500000e+00   1.0  1.520270e+00  2.295198e-01
C(Fertilizer):C(Sunlight)  4.733165e-29   2.0  4.797127e-30  1.000000e+00
Residual                   1.184000e+02  24.0           NaN           NaN
Reject null hypothesis: The means are not equal, as the p-value: {p_val} is less than 0.05


C:\Users\Hp\AppData\Local\Temp\ipykernel_252\1127512014.py:25: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  if anova_table["PR(>F)"][0] < 0.05:


The above code is performing a two-way ANOVA (Analysis of Variance) test using the statsmodels library on the growth of plants with `three` different types of `fertilizers` and `two` different `sunlight conditions`. Here's a step-by-step explanation:

1. `Import the necessary libraries:` pandas, statsmodels.api, and statsmodels.formula.api.
2. `Define the sample data` for the growth of plants with three types of fertilizers and two sunlight conditions and create a DataFrame.
3. `Fit the model` using the ols (Ordinary Least Squares) method from statsmodels.formula.api. The formula includes the main effects (Fertilizer and Sunlight) and the interaction effect (Fertilizer:Sunlight).
4. Perform the ANOVA using `anova_lm()` function from statsmodels.api and print the summary table.
5. Based on the p-value, a decision is made about the null hypothesis. If the p-value is less than 0.05, we reject the null hypothesis and conclude that at least one of the means of the plant growth for the three fertilizers or the two sunlight conditions is different. If the p-value is greater than or equal to 0.05, we fail to reject the null hypothesis and conclude that all the means are equal.

**Observations from the output:**

- The output shows that the `F-statistic` and the `p-value` for the Fertilizer effect are significant (`p-value < 0.05`), so we `reject the null hypothesis` for the Fertilizer effect. 
- This suggests that at least one of the means of the plant growth for the three fertilizers is `different` according to the two-way ANOVA test.

**Note:** The warning message indicates that using integer keys to access a value in a pandas Series is deprecated and will be removed in a future version. To access a value by position, you should use the `iloc` property. So, you should replace `anova_table["PR(>F)"][0]` with `anova_table["PR(>F)"].iloc[0]`.

## **TIPS:**
- For `One-Way ANOVA`, if the p-value is less than 0.05, it suggests a significant difference in means among the groups.
- For `Two-Way ANOVA`, we look at the p-values for each factor and their interaction. A p-value less than 0.05 indicates a significant effect.

These examples should give you a good starting point for conducting ANOVA analyses in Python. Remember, the interpretation of your results should always take into account the context of your data and the specific question you are trying to answer.

## **3. N-way ANOVA:**

> `N-way ANOVA`, also known as `factorial ANOVA`, is used when you have more than two independent variables. It allows you to analyze the effects of each factor on the dependent variable and the interaction effects between factors.

**Example: Three-Way ANOVA**

Suppose we have an experimental data set with three factors:

1. Fertilizer Type (3 levels: F1, F2, F3)
2. Sunlight Exposure (2 levels: High, Low)
3. Watering Frequency (2 levels: Regular, Sparse)

**We want to study the impact of these factors and their interactions on plant growth.**

In [12]:
# importing libraries
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols

# Sample data
data = pd.DataFrame({
    "Growth": [20, 22, 19, 24, 25, 28, 30, 27, 26, 29, 18, 20, 22, 19, 24,
               21, 23, 20, 25, 26, 29, 31, 28, 27, 30, 19, 21, 23, 20, 25,
               20, 22, 21, 23, 24, 26, 28, 25, 27, 29, 17, 19, 21, 18, 20],
    "Fertilizer": ["F1", "F1", "F1", "F1", "F1", "F2", "F2", "F2", "F2", "F2", 
                   "F3", "F3", "F3", "F3", "F3", "F1", "F1", "F1", "F1", "F1", 
                   "F2", "F2", "F2", "F2", "F2", "F3", "F3", "F3", "F3", "F3",
                   "F1", "F1", "F1", "F1", "F1", "F2", "F2", "F2", "F2", "F2", 
                   "F3", "F3", "F3", "F3", "F3"],
    "Sunlight": ["High", "High", "High", "High", "High", "High", "High", "High", "High", "High", 
                 "High", "High", "High", "High", "High", "Low", "Low", "Low", "Low", "Low", 
                 "Low", "Low", "Low", "Low", "Low", "Low", "Low", "Low", "Low", "Low",
                 "High", "High", "High", "High", "High", "High", "High", "High", "High", "High", 
                 "High", "High", "High", "High", "High"],
    "Watering": ["Regular", "Regular", "Regular", "Regular", "Regular", 
                 "Regular", "Regular", "Regular", "Regular", "Regular",
                 "Regular", "Regular", "Regular", "Regular", "Regular",
                 "Sparse", "Sparse", "Sparse", "Sparse", "Sparse", 
                 "Sparse", "Sparse", "Sparse", "Sparse", "Sparse",
                 "Sparse", "Sparse", "Sparse", "Sparse", "Sparse",
                 "Regular", "Regular", "Regular", "Regular", "Regular", 
                 "Regular", "Regular", "Regular", "Regular", "Regular",
                 "Regular", "Regular", "Regular", "Regular", "Regular"]
})

# Fit the model
model = ols('Growth ~ C(Fertilizer) + C(Sunlight) + C(Watering) + C(Fertilizer):C(Sunlight) + C(Fertilizer):C(Watering) + C(Sunlight):C(Watering) + C(Fertilizer):C(Sunlight):C(Watering)', data=data).fit()

# Perform three-way ANOVA
anova_results = sm.stats.anova_lm(model, typ=2)

print(anova_results)


# print the results based on if the p-value is less than 0.05

if anova_results["PR(>F)"][0] < 0.05:
    print("Reject null hypothesis: The means are not equal, as the p-value: {p_val} is less than 0.05")
else:
    print("Fail to reject null hypothesis: The means are equal, as the p-value: {p_val} is greater than 0.05")


                                             sum_sq    df             F  \
C(Fertilizer)                          5.672800e-02   2.0  7.032396e-03   
C(Sunlight)                           -5.817865e-12   1.0 -1.442446e-12   
C(Watering)                            3.467164e-13   1.0  8.596273e-14   
C(Fertilizer):C(Sunlight)             -1.058405e-12   2.0 -1.312072e-13   
C(Fertilizer):C(Watering)              1.248386e-12   2.0  1.547585e-13   
C(Sunlight):C(Watering)                2.054444e+01   1.0  5.093664e+00   
C(Fertilizer):C(Sunlight):C(Watering)  1.088889e+00   2.0  1.349862e-01   
Residual                               1.573000e+02  39.0           NaN   

                                         PR(>F)  
C(Fertilizer)                          0.933597  
C(Sunlight)                            1.000000  
C(Watering)                            1.000000  
C(Fertilizer):C(Sunlight)              1.000000  
C(Fertilizer):C(Watering)              1.000000  
C(Sunlight):C(Watering) 

c:\Users\aammar\miniconda3\envs\python_eda\Lib\site-packages\statsmodels\base\model.py:1888: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 2, but rank is 1
  warnings.warn('covariance of constraints does not have full '
c:\Users\aammar\miniconda3\envs\python_eda\Lib\site-packages\statsmodels\base\model.py:1888: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 2, but rank is 1
  warnings.warn('covariance of constraints does not have full '
C:\Users\aammar\AppData\Local\Temp\ipykernel_8868\2024720544.py:42: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  if anova_results["PR(>F)"][0] < 0.05:


The above code is performing a three-way ANOVA (Analysis of Variance) test using the statsmodels library on the growth of plants with `three` different types of `fertilizers`, `two` different `sunlight conditions`, and `two` different `watering conditions`. Here's a step-by-step explanation:

1. `Import the necessary libraries:` pandas, statsmodels.api, and statsmodels.formula.api.
2. `Define the sample data` for the growth of plants with three types of fertilizers, two sunlight conditions, and two watering conditions and create a DataFrame.
3. `Fit the model` using the ols (Ordinary Least Squares) method from statsmodels.formula.api. The formula includes the main effects (Fertilizer, Sunlight, Watering) and the interaction effects (Fertilizer:Sunlight, Fertilizer:Watering, Sunlight:Watering, Fertilizer:Sunlight:Watering).
4. Perform the `three-way ANOVA` using `anova_lm()` function from statsmodels.api and print the summary table.
5. Based on the p-value, a decision is made about the null hypothesis. If the p-value is less than 0.05, we reject the null hypothesis and conclude that at least one of the means of the plant growth for the three fertilizers, the two sunlight conditions, or the two watering conditions is different. If the p-value is greater than or equal to 0.05, we fail to reject the null hypothesis and conclude that all the means are equal.

**Observations from the output:**

- The output shows that the `F-statistic` and the `p-value` for the `Sunlight:Watering` effect are significant (`p-value < 0.05`), so we `reject the null hypothesis` for the Sunlight:Watering effect. 
- This suggests that the interaction between sunlight and watering has a significant effect on plant growth according to the three-way ANOVA test.

**Note:** The warning messages indicate that the covariance of constraints does not have full rank, which means that there might be multicollinearity in the data. The FutureWarning message indicates that using integer keys to access a value in a pandas Series is deprecated and will be removed in a future version. To access a value by position, you should use the `iloc` property. So, you should replace `anova_results["PR(>F)"][0]` with `anova_results["PR(>F)"].iloc[0]`.

### **TIPS:**
In the output, you'll see p-values for:

- The main effects of each factor (Fertilizer, Sunlight, Watering)
- The interaction effects between two factors (e.g., Fertilizer:Sunlight)
- The interaction effect among all three factors (Fertilizer:Sunlight:Watering)

A p-value less than 0.05 typically suggests a statistically significant effect. However, interpreting ANOVA results can be complex, especially with interactions. You should consider the practical significance and the context of your experiment alongside the statistical results.

Remember, ANOVA makes certain assumptions (normality, homogeneity of variance, and independence), which should be tested before running the analysis.

---
# **4. Post-hoc Tests**

> A post-hoc test is a statistical analysis conducted after an initial test like ANOVA. 
> It's used to determine `which specific groups' means (average values) are different from each other`. 
> Post-hoc tests adjust for multiple comparisons to control the probability of falsely rejecting the null hypothesis (Type I error).

Common post-hoc tests include: 
1. Tukey's HSD
2. Bonferroni correction
3. Scheffé's method

- After conducting an ANOVA and finding a significant difference, a post hoc test is needed to determine exactly which groups differ from each other.  

#### **`Important Considerations:`**
- **Interpreting Results:** Tukey's test results will show pairwise comparisons between groups and indicate which differences are statistically significant.
- **Alpha Level:** The alpha parameter is typically set at 0.05, which is the threshold for significance.
- **Assumptions:** Like ANOVA, Tukey's HSD assumes homogeneity of variances and normally distributed data within each group.

Here, I'll demonstrate post hoc tests for one-way, two-way, and N-way ANOVA using Python.

## **4.1 Post-hoc Tests for `One-Way ANOVA` using Tukey's HSD:**

In [15]:
# Import libraries
from statsmodels.stats.multicomp import pairwise_tukeyhsd
import numpy as np

# Sample data
data = {
    'Growth': np.concatenate([fertilizer1, fertilizer2, fertilizer3]),
    'Fertilizer': ['F1']*len(fertilizer1) + ['F2']*len(fertilizer2) + ['F3']*len(fertilizer3)
}

# Convert to DataFrame
df = pd.DataFrame(data)

# Tukey's HSD test
tukey = pairwise_tukeyhsd(endog=df['Growth'], groups=df['Fertilizer'], alpha=0.05) # alpha is the confidence level (95%)
print(tukey)

 Multiple Comparison of Means - Tukey HSD, FWER=0.05 
group1 group2 meandiff p-adj   lower    upper  reject
-----------------------------------------------------
    F1     F2      6.0 0.0029   2.2523  9.7477   True
    F1     F3     -1.4 0.5928  -5.1477  2.3477  False
    F2     F3     -7.4 0.0005 -11.1477 -3.6523   True
-----------------------------------------------------


The code is performing Tukey's HSD (Honestly Significant Difference) post-hoc test using the statsmodels library on the `growth` of plants with `three different types of fertilizers`. Here's a step-by-step explanation:

1. `Import the necessary libraries:` statsmodels.stats.multicomp and numpy.
2. `Define the sample data` for the growth of plants with three types of fertilizers and create a DataFrame.
3. `Perform Tukey's HSD test` using `pairwise_tukeyhsd()` function from statsmodels.stats.multicomp. This function returns a result object that can be printed to get a summary of the test.
4. Print the `summary` of the test.

**Observations from the output:**

- The output shows the `pairwise comparison`s between the three fertilizers. 
- For each pair, it shows the `difference in means` (meandiff), the adjusted p-value (p-adj), the lower and upper bounds of the 95% confidence interval for the difference in means, and whether to reject the null hypothesis (reject) at the 5% significance level.

- The results suggest that there is a `significant difference` in plant growth between Fertilizer 1 and Fertilizer 2 (`p-adj = 0.0029`, reject = True)
- There is a `significant differene` between Fertilizer 2 and Fertilizer 3 (`p-adj = 0.0005`, reject = True). 
- There is `no significant difference` in plant growth between Fertilizer 1 and Fertilizer 3 (`p-adj = 0.5928`, reject = False).

## **4.2 Post-hoc Tests for `Two-Way ANOVA` using Tukey's HSD:**

In [9]:
from statsmodels.stats.multicomp import pairwise_tukeyhsd
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols

# Sample data
data = pd.DataFrame({
    "Growth": [20, 22, 19, 24, 25, 28, 30, 27, 26, 29, 18, 20, 22, 19, 24,
               21, 23, 20, 25, 26, 29, 31, 28, 27, 30, 19, 21, 23, 20, 25],
    "Fertilizer": ["F1", "F1", "F1", "F1", "F1", "F2", "F2", "F2", "F2", "F2", 
                   "F3", "F3", "F3", "F3", "F3", "F1", "F1", "F1", "F1", "F1", 
                   "F2", "F2", "F2", "F2", "F2", "F3", "F3", "F3", "F3", "F3"],
    "Sunlight": ["High", "High", "High", "High", "High", "High", "High", "High", "High", "High", 
                 "High", "High", "High", "High", "High", "Low", "Low", "Low", "Low", "Low", 
                 "Low", "Low", "Low", "Low", "Low", "Low", "Low", "Low", "Low", "Low"]
})

tukey = pairwise_tukeyhsd(data['Growth'], data['Fertilizer'] + data['Sunlight'], alpha=0.05)
print(tukey)

 Multiple Comparison of Means - Tukey HSD, FWER=0.05 
group1 group2 meandiff p-adj   lower    upper  reject
-----------------------------------------------------
F1High  F1Low      1.0 0.9786  -3.3434  5.3434  False
F1High F2High      6.0 0.0032   1.6566 10.3434   True
F1High  F2Low      7.0 0.0006   2.6566 11.3434   True
F1High F3High     -1.4 0.9145  -5.7434  2.9434  False
F1High  F3Low     -0.4 0.9997  -4.7434  3.9434  False
 F1Low F2High      5.0 0.0176   0.6566  9.3434   True
 F1Low  F2Low      6.0 0.0032   1.6566 10.3434   True
 F1Low F3High     -2.4 0.5396  -6.7434  1.9434  False
 F1Low  F3Low     -1.4 0.9145  -5.7434  2.9434  False
F2High  F2Low      1.0 0.9786  -3.3434  5.3434  False
F2High F3High     -7.4 0.0003 -11.7434 -3.0566   True
F2High  F3Low     -6.4 0.0016 -10.7434 -2.0566   True
 F2Low F3High     -8.4    0.0 -12.7434 -4.0566   True
 F2Low  F3Low     -7.4 0.0003 -11.7434 -3.0566   True
F3High  F3Low      1.0 0.9786  -3.3434  5.3434  False
----------------------------

The above code is performing Tukey's HSD (Honestly Significant Difference) post-hoc test using the statsmodels library on the growth of plants with `three` different types of `fertilizers` and `two` different `sunlight conditions` (High and Low). Here's a step-by-step explanation:

1. `Import the necessary libraries:` statsmodels.stats.multicomp, pandas, statsmodels.api, and statsmodels.formula.api.
2. `Define the sample data` for the growth of plants with three types of fertilizers and two sunlight conditions and create a DataFrame.
3. `Perform Tukey's HSD test` using `pairwise_tukeyhsd()` function from statsmodels.stats.multicomp. This function returns a result object that can be printed to get a summary of the test.
4. Print the `summary` of the test.

**Observations from the output:**

- The output shows the `pairwise comparisons` between the `six groups` (F1High, F1Low, F2High, F2Low, F3High, F3Low). 
- For each pair, it shows the difference in means (`meandiff`), the adjusted p-value (`p-adj`), the `lower` and `upper bounds` of the 95% confidence interval for the difference in means, and whether to reject the null hypothesis (reject) at the 5% significance level.

- The results suggest that there are `significant differences` in plant growth between several pairs of groups. 
- For example
- There is a `significant difference` in Plant growth between: 
  - `F1High` and `F2High` (`p-adj = 0.0032`, reject = True)
  - `F1High` and `F2Low` (`p-adj = 0.0006`, reject = True)
  - `F2High` and `F3High`, `F2High` and `F3Low`, `F2Low` and `F3High`, and `F2Low` and `F3Low`. 
- There are no significant differences in plant growth between the other pairs of groups.

## **4.3 Post-hoc Tests for `N-Way ANOVA` using Tukey's HSD:**

In [10]:
from statsmodels.stats.multicomp import pairwise_tukeyhsd
# Sample data
data = pd.DataFrame({
    "Growth": [20, 22, 19, 24, 25, 28, 30, 27, 26, 29, 18, 20, 22, 19, 24,
               21, 23, 20, 25, 26, 29, 31, 28, 27, 30, 19, 21, 23, 20, 25,
               20, 22, 21, 23, 24, 26, 28, 25, 27, 29, 17, 19, 21, 18, 20],
    "Fertilizer": ["F1", "F1", "F1", "F1", "F1", "F2", "F2", "F2", "F2", "F2", 
                   "F3", "F3", "F3", "F3", "F3", "F1", "F1", "F1", "F1", "F1", 
                   "F2", "F2", "F2", "F2", "F2", "F3", "F3", "F3", "F3", "F3",
                   "F1", "F1", "F1", "F1", "F1", "F2", "F2", "F2", "F2", "F2", 
                   "F3", "F3", "F3", "F3", "F3"],
    "Sunlight": ["High", "High", "High", "High", "High", "High", "High", "High", "High", "High", 
                 "High", "High", "High", "High", "High", "Low", "Low", "Low", "Low", "Low", 
                 "Low", "Low", "Low", "Low", "Low", "Low", "Low", "Low", "Low", "Low",
                 "High", "High", "High", "High", "High", "High", "High", "High", "High", "High", 
                 "High", "High", "High", "High", "High"],
    "Watering": ["Regular", "Regular", "Regular", "Regular", "Regular", 
                 "Regular", "Regular", "Regular", "Regular", "Regular",
                 "Regular", "Regular", "Regular", "Regular", "Regular",
                 "Sparse", "Sparse", "Sparse", "Sparse", "Sparse", 
                 "Sparse", "Sparse", "Sparse", "Sparse", "Sparse",
                 "Sparse", "Sparse", "Sparse", "Sparse", "Sparse",
                 "Regular", "Regular", "Regular", "Regular", "Regular", 
                 "Regular", "Regular", "Regular", "Regular", "Regular",
                 "Regular", "Regular", "Regular", "Regular", "Regular"]
})

tukey = pairwise_tukeyhsd(data['Growth'], data['Fertilizer'] + data['Sunlight'] + data['Watering'], alpha=0.05)
print(tukey)

        Multiple Comparison of Means - Tukey HSD, FWER=0.05        
    group1        group2    meandiff p-adj   lower    upper  reject
-------------------------------------------------------------------
F1HighRegular   F1LowSparse      1.0 0.9419  -2.2956  4.2956  False
F1HighRegular F2HighRegular      5.5    0.0   2.8092  8.1908   True
F1HighRegular   F2LowSparse      7.0    0.0   3.7044 10.2956   True
F1HighRegular F3HighRegular     -2.2 0.1647  -4.8908  0.4908  False
F1HighRegular   F3LowSparse     -0.4 0.9991  -3.6956  2.8956  False
  F1LowSparse F2HighRegular      4.5 0.0027   1.2044  7.7956   True
  F1LowSparse   F2LowSparse      6.0 0.0004   2.1946  9.8054   True
  F1LowSparse F3HighRegular     -3.2 0.0613  -6.4956  0.0956  False
  F1LowSparse   F3LowSparse     -1.4 0.8775  -5.2054  2.4054  False
F2HighRegular   F2LowSparse      1.5 0.7478  -1.7956  4.7956  False
F2HighRegular F3HighRegular     -7.7    0.0 -10.3908 -5.0092   True
F2HighRegular   F3LowSparse     -5.9 0.0001  -9.

The above code is performing Tukey's HSD (Honestly Significant Difference) post-hoc test using the statsmodels library on the growth of plants with three different types of fertilizers, two different sunlight conditions (High and Low), and two different watering conditions (Regular and Sparse). Here's a step-by-step explanation:

1. Import the necessary library: `pairwise_tukeyhsd` from `statsmodels.stats.multicomp`.
2. Define the sample data for the growth of plants with three types of fertilizers, two sunlight conditions, and two watering conditions, and create a DataFrame.
3. Perform Tukey's HSD test using `pairwise_tukeyhsd()` function from statsmodels.stats.multicomp. This function returns a result object that can be printed to get a summary of the test.
4. Print the summary of the test.

**Observations from the output:**

- There is a `significant difference` in plant growth between: 
  - F1HighRegular and F2HighRegular (p-adj = 0.0, reject = True)
  - F1HighRegular and F2LowSparse (p-adj = 0.0, reject = True)
  - F1LowSparse and F2HighRegular (p-adj = 0.0027, reject = True)
  - F1LowSparse and F2LowSparse (p-adj = 0.0004, reject = True)
  - F2HighRegular and F3HighRegular (p-adj = 0.0, reject = True)
  - F2HighRegular and F3LowSparse (p-adj = 0.0001, reject = True)
  - F2LowSparse and F3HighRegular (p-adj = 0.0, reject = True)
  - F2LowSparse and F3LowSparse (p-adj = 0.0, reject = True)

- There are `no significant differences` in plant growth between the other pairs of groups (where reject = False).

These results suggest that the type of fertilizer, sunlight condition, and watering condition all have a significant impact on plant growth.

`These examples provide a basic framework for conducting post hoc analyses in Python. The specific methods and interpretations will depend on your data and the context of your research.`

---
## **4.4. Bonferri Correction**

- The Bonferroni test is a `type of Post Hoc Analysis` used after conducting ANOVA when multiple pairwise comparisons are needed. 
- It's a conservative method that adjusts the significance level to account for the increased risk of Type I errors (false positives) due to multiple testing. 
- The Bonferroni correction simply divides the alpha level by the number of comparisons.

In [11]:
import scipy.stats as stats
import pandas as pd

# Sample data
fertilizer1 = [20, 22, 19, 24, 25]
fertilizer2 = [28, 30, 27, 26, 29]
fertilizer3 = [18, 20, 22, 19, 24]

# Create DataFrame
data = {
    'Growth': fertilizer1 + fertilizer2 + fertilizer3,
    'Fertilizer': ['F1']*len(fertilizer1) + ['F2']*len(fertilizer2) + ['F3']*len(fertilizer3)
}
df = pd.DataFrame(data)

# Number of comparisons
num_comparisons = 3 # For 3 groups, we have 3 pairwise comparisons (F1 vs F2, F1 vs F3, F2 vs F3)

# Adjusted alpha level (for significance)
alpha = 0.05 / num_comparisons

# Conduct pairwise t-tests with Bonferroni correction
pairwise_results = []
for group1 in df['Fertilizer'].unique():
    for group2 in df['Fertilizer'].unique():
        if group1 < group2: # To avoid duplicate comparisons
            group1_data = df[df['Fertilizer'] == group1]['Growth']
            group2_data = df[df['Fertilizer'] == group2]['Growth']
            t_stat, p_val = stats.ttest_ind(group1_data, group2_data)
            p_val_adjusted = p_val * num_comparisons
            pairwise_results.append((f'{group1} vs {group2}', t_stat, p_val_adjusted))

# Print results
for result in pairwise_results:
    group_comparison, t_stat, p_val_adjusted = result
    print(f"{group_comparison}: t-statistic = {t_stat:.3f}, p-value (adjusted) = {p_val_adjusted:.3f}")

F1 vs F2: t-statistic = -4.472, p-value (adjusted) = 0.006
F1 vs F3: t-statistic = 0.893, p-value (adjusted) = 1.194
F2 vs F3: t-statistic = 5.744, p-value (adjusted) = 0.001


The above code is performing `pairwise t-tests` with Bonferroni correction on the growth of plants with `three` different types of fertilizers. Here's a step-by-step explanation:

1. `Import the necessary libraries`: scipy.stats and pandas.
2. `Define the sample data` for the growth of plants with three types of fertilizers and create a DataFrame.
3. Calculate the `number of comparisons` and the adjusted `alpha level` for significance.
4. Perform pairwise t-tests with Bonferroni correction. 
   - For each pair of groups (F1 vs F2, F1 vs F3, F2 vs F3), it calculates the t-statistic and the p-value, adjusts the p-value for multiple comparisons, and stores the results.
5. Print the results.

**Observations from the output:**

- There is a `significant difference` in plant growth between `F1` and `F2` (t-statistic = -4.472, adjusted p-value = 0.006). 
  - The negative t-statistic indicates that the mean growth for F1 is less than the mean growth for F2.
- There is a `significant difference` in plant growth between `F2` and `F3` (t-statistic = 5.744, adjusted p-value = 0.001). 
  - The positive t-statistic indicates that the mean growth for F2 is greater than the mean growth for F3.
- There is `no significant difference` in plant growth between `F1` and `F3 `(t-statistic = 0.893, adjusted p-value = 1.194). 
  - The adjusted p-value is greater than 0.05, so we fail to reject the null hypothesis of equal means.

These results suggest that the type of fertilizer has a significant impact on plant growth. Specifically, F2 appears to promote more growth than F1 and F3.

### **TIPS**
- Each pairwise comparison is adjusted for multiple testing.
- A significant result (adjusted p-value < 0.05) indicates a significant difference between those groups.
- This method is conservative and reduces the chance of Type I errors but increases the chance of Type II errors (false negatives).
### **Notes**
- The Bonferroni test is appropriate when you have a small number of comparisons. With a large number of comparisons, it becomes overly conservative.
- The method assumes independent tests and may not be suitable for all types of data or comparisons.

# **Types of PostHoc Tests with their pros and cons.**

There are several post hoc tests used in statistics, each with its own strengths and weaknesses. Here's a table summarizing some of the most common tests:

| Post Hoc Test        | Pros                                                    | Cons                                                    | Ideal Usage                                           |
|----------------------|---------------------------------------------------------|---------------------------------------------------------|-------------------------------------------------------|
| **Tukey's HSD**      | - Controls Type I error well<br>- Comparatively robust  | - Can be conservative<br>- Less powerful for unequal sample sizes | When equal sample sizes and normally distributed data are assumed. |
| **Bonferroni**       | - Simple to compute<br>- Very conservative              | - Increases Type II errors<br>- Can be too stringent for many comparisons | When few comparisons are made; useful in controlling Type I error in multiple testing. |
| **Scheffé's Test**   | - Flexible for any number of comparisons                | - More conservative than others<br>- Can be less powerful | When flexibility in hypothesis testing after ANOVA is needed; good for complex designs. |
| **Dunn's Test**      | - Suitable for non-parametric data                      | - Less powerful than parametric tests<br>- Multiple comparison adjustments can be complex | When data do not meet parametric assumptions, particularly with Kruskal-Wallis test. |
| **Holm's Method**    | - Less conservative than Bonferroni<br>- Controls Type I error well | - More complex calculation<br>- Can still be conservative | When a balance between Type I error control and power is needed, especially with multiple comparisons. |
| **Fisher's LSD**     | - More powerful (higher chance to detect real differences) | - Higher risk of Type I error<br>- Not recommended when there are many comparisons | When comparisons are planned and limited, often used in exploratory data analysis. |
| **Ryan-Einot-Gabriel-Welsch Q (REGWQ)** | - Controls the error rate well<br>- Good for unequal sample sizes | - Complex calculation<br>- Can be conservative | When there are unequal sample sizes, and control over Type I error is important. |
| **Newman-Keuls**     | - More powerful for detecting differences                | - Higher risk of Type I errors than other methods<br>- Not recommended for many comparisons | When sample sizes are equal and the data are normally distributed; less used due to higher error risks. |

### **Notes:**
1. **Choice of Test**: The choice of post hoc test largely depends on the nature of your data, the number of comparisons, and the balance you want to strike between the risks of Type I and Type II errors.
2. **Data Assumptions**: Some tests assume normally distributed data and equal variances, while others are non-parametric and do not make these assumptions.
3. **Type I and II Errors**: There's often a trade-off between the risk of Type I errors (false positives) and Type II errors (false negatives). More conservative tests (like Bonferroni) reduce the risk of Type I errors but increase the risk of Type II errors.

When conducting post hoc tests, it's essential to understand these pros and cons to choose the most appropriate test for your specific statistical analysis.

---